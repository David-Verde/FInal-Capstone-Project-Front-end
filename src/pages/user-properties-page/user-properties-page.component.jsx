/* eslint-disable camelcase */
import pluralize from 'pluralize';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { CustomSelector } from '../../components/custom-selector/custom-selector.component';
import { PropertiesGrid } from '../../components/properties-grid/properties-grid.component';
import { SectionTitle } from '../../components/section-title/section-title.component';
import { SimpleCard } from '../../components/simple-card/simple-card.component';
import { fetchUserPropertiesAsync } from '../../redux/property/property.actions';
import { selectUserProperties } from '../../redux/property/property.selectors';
import { selectCurrentUser } from '../../redux/user/user.selectors';

import './user-properties-page.styles.scss';

const PROPERTY_OPTIONS = {
  ALL: 'all',
  PRIVATE: 'private',
  PUBLIC: 'public',
};

const options = [
  {
    name: 'All',
    value: PROPERTY_OPTIONS.ALL,
  },
  {
    name: 'Private',
    value: PROPERTY_OPTIONS.PRIVATE,
  },
  {
    name: 'Public',
    value: PROPERTY_OPTIONS.PUBLIC,
  },
];

export function UserPropertiesPage() {
  const dispatch = useDispatch();
  const currentUser = useSelector(selectCurrentUser);
  const userProperties = useSelector(selectUserProperties);
  const [propertiesToShow, setPropertiesToShow] = useState([]);

  useEffect(() => {
    if (currentUser) {
      dispatch(fetchUserPropertiesAsync(currentUser.id));
    }
  }, [currentUser]);

  useEffect(() => {
    setPropertiesToShow(userProperties);
  }, [userProperties]);

  const handleSelectOption = (option) => {
    switch (option) {
      case PROPERTY_OPTIONS.PRIVATE:
        setPropertiesToShow(
          userProperties.filter(({ is_public }) => !is_public),
        );
        break;
      case PROPERTY_OPTIONS.PUBLIC:
        setPropertiesToShow(
          userProperties.filter(({ is_public }) => is_public),
        );
        break;
      default:
        setPropertiesToShow(userProperties);
        break;
    }
  };

  return (
    <section className="user-properties">
      <SimpleCard>
        <header className="user-properties__header">
          <SectionTitle
            title={pluralize('Property', propertiesToShow.length, true)}
          />
          <CustomSelector
            handleChangeCallback={handleSelectOption}
            defaultOption={PROPERTY_OPTIONS.ALL}
            options={options}
          />
        </header>
      </SimpleCard>

      <div className="user-properties__content">
        <PropertiesGrid properties={propertiesToShow} isListing={false} />
      </div>
    </section>
  );
}
