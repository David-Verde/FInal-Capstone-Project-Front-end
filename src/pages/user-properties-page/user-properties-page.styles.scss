.user-properties {
  padding-top: 2.4rem;

  &__header {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  &__content {
    margin-top: 3.2rem;
    padding: 0 2.4rem;

    @media only screen and (min-width: 48em) {
      padding: 0;
    }
  }

  &__text {
    width: 100%;
    font-size: 1.6rem;
    font-weight: 700;
    text-align: center;
    padding: 1.6rem 0 4.8rem;
  }
}
