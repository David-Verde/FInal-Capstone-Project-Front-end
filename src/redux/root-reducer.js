import categoryReducer from './category/category.reducer';
import propertyReducer from './property/property.reducer';
import reservationReducer from './reservation/reservation.reducer';
import statusReducer from './status/status.reducer';
import userReducer from './user/user.reducer';

export const rootReducer = {
  user: userReducer,
  property: propertyReducer,
  category: categoryReducer,
  reservation: reservationReducer,
  status: statusReducer,
};
