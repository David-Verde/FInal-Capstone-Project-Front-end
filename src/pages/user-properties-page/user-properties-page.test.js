import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { store } from '../../redux/store';
import { UserPropertiesPage } from './user-properties-page.component';

const renderWithProviders = (reduxStore, route) => render(
  <Provider store={reduxStore}>
    <MemoryRouter initialEntries={[route]}>
      <UserPropertiesPage />
    </MemoryRouter>
  </Provider>,
);

describe('UserPropertiesPage component', () => {
  it('should render correctly', () => {
    const component = renderWithProviders(store, '/');
    expect(component).toMatchSnapshot();
  });
});
