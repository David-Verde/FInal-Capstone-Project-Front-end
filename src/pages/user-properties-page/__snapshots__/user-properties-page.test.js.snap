// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPropertiesPage component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="user-properties"
      >
        <div
          class="simple-card"
        >
          <header
            class="user-properties__header"
          >
            <h1
              class="section-title"
            >
              0 Properties
            </h1>
            <select
              class="custom-selector"
              style="width: 94px;"
            >
              <option
                selected=""
                value="all"
              >
                All
              </option>
              <option
                value="private"
              >
                Private
              </option>
              <option
                value="public"
              >
                Public
              </option>
            </select>
          </header>
        </div>
        <div
          class="user-properties__content"
        >
          <p
            class="user-properties__text"
          >
            —  No properties  —
          </p>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="user-properties"
    >
      <div
        class="simple-card"
      >
        <header
          class="user-properties__header"
        >
          <h1
            class="section-title"
          >
            0 Properties
          </h1>
          <select
            class="custom-selector"
            style="width: 94px;"
          >
            <option
              selected=""
              value="all"
            >
              All
            </option>
            <option
              value="private"
            >
              Private
            </option>
            <option
              value="public"
            >
              Public
            </option>
          </select>
        </header>
      </div>
      <div
        class="user-properties__content"
      >
        <p
          class="user-properties__text"
        >
          —  No properties  —
        </p>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
